aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -428,6 +428,39 @@ Tensor matmul(
     Tensor output = has_out ? at::_unsafe_view(at::mm_out(out, t1, t2), output_size)
                             : at::_unsafe_view(t1.mm(t2), output_size);
     return has_out ? out.set_(output) : output;
+  } else if ((dim_tensor1 == 1 || dim_tensor1 == 2) && dim_tensor2 >= 3) {
+    // optimization: transpose the inner dimensions of the arguments, call
+    // matmul on the swapped arguments, then transpose the inner dimensions
+    // of the result.
+    const int64_t n = dim_tensor1 == 2 ? tensor1.size(-2) : 1;
+    const int64_t m = tensor1.size(-1);
+    const int64_t p = tensor2.size(-1);
+
+    std::vector<int64_t> perm(dim_tensor2);
+    for (int64_t i = 0; i < dim_tensor2-2; i++) {
+        perm[i] = i;
+    }
+    perm[dim_tensor2-2] = dim_tensor2-1;
+    perm[dim_tensor2-1] = dim_tensor2-2;
+
+    const Tensor t2_T = tensor2.permute(perm);
+    const Tensor t1_T = dim_tensor1 == 2 ? tensor1.t() : tensor1.reshape({n, m}).t();
+    const Tensor res_T = matmul(out_opt, t2_T, t1_T);
+
+    if (dim_tensor1 == 2) {
+      Tensor res = res_T.permute(perm).contiguous();
+      return has_out ? out.set_(res) : res;
+    }
+    else {
+      const IntArrayRef sizes = tensor2.sizes();
```

Comment:
This is fine, but you can also write it as:

```
std::vector<int64_t> shape = tensor2.sizes().slice(0, dim_tensor2 - 2).vec();
shape.push_back(p);
```

Reply:
Thanks, that looks better, too.